src/TCommandLine.cpp

```diff
@@ -115,85 +193,198 @@ bool TCommandLine::event(QEvent* event)
-                }
+                // If Ctrl-Return is pressed - scroll to the bottom of text:
+                mpConsole->mLowerPane->mCursorY = mpConsole->buffer.size();
+                mpConsole->mLowerPane->hide();
+                mpConsole->buffer.mCursorY = mpConsole->buffer.size();
+                mpConsole->mUpperPane->mCursorY = mpConsole->buffer.size();
+                mpConsole->mUpperPane->mIsTailMode = true;
+                mpConsole->mUpperPane->updateScreenView();
+                mpConsole->mUpperPane->forceUpdate();
                 ke->accept();
                 return true;
-
             }
+            break;
+        case Qt::Key_PageUp:
+            if ((ke->modifiers() & allModifiers) == Qt::NoModifier) {
+                mpConsole->scrollUp(mpHost->mScreenHeight);
+                ke->accept();
+                return true;
+            }
+            break;
 
+        case Qt::Key_PageDown:
             if ((ke->modifiers() & allModifiers) == Qt::NoModifier) {
-                handleTabCompletion(true);
+                mpConsole->scrollDown(mpHost->mScreenHeight);
                 ke->accept();
                 return true;
+            }
+            break;
+        case Qt::Key_C:
+            if (((ke->modifiers() & allModifiers) == Qt::ControlModifier)
+                && (mpConsole->mUpperPane->mSelectedRegion != QRegion(0, 0, 0, 0))) {
+
+                // Only process as a Control-C if it is EXACTLY those two keys
+                // and no other AND there is a selection active in the TConsole
+                mpConsole->mUpperPane->slot_copySelectionToClipboard();
+                ke->accept();
+                return true;
+
+            }
+            break;
+        case Qt::Key_V:
+            if (mpHost->mEnableReadline && ((ke->modifiers() & allModifiers) == Qt::ControlModifier)) {
+                auto clipboard = QApplication::clipboard();
+                QTextCursor c = textCursor();
+                QString text = clipboard->text();
+                // GNU readline is about reading _lines_, so pasting endlines
+                // leads to trouble.
+                text = text.replace('\n', '\t');
+                c.insertText(text);
+                ke->accept();
+                adjustHeight();
+                return true;
+
+            }
+            break;
+        }
+        // Missing: C-x for cut, tab completion using GNU readline
+
+        if (mpHost->mEnableReadline) {
+            if (keyText.length() == 0)
+                return false;
+            if (keyText.length() == 1 && (ke->modifiers() & Qt::AltModifier)){
+                keyText = "\x1b" + keyText;
+            }
+            mTabCompletionCount = -1;
+            ke->accept();
+
+            switch_rl_here();
+
+            // Copy the widget's view to readline's view.
+
+            // In order to figure out where the current cursor position is
+            // in the UTF-8 encoded string, we separately encode the
```

Comment:
The widget's text is NOT *UTF-8* encoded - it is *UTF-**16*** - does that break things here?

Reply:
See my longer comment: this is referring to the encoding _for readline_.  We want to convert the QTextCursor position (which is a number of characters) into the rl_line_buffer (which is a number of bytes in the often-but-not-necessarily-utf8 string).